Libs/MRML/Core/vtkMRMLNodePropertyMacros.h

```diff
@@ -0,0 +1,253 @@
+/*=auto=========================================================================
+
+  Portions (c) Copyright 2005 Brigham and Women's Hospital (BWH) All Rights Reserved.
+
+  See COPYRIGHT.txt
+  or http://www.slicer.org/copyright/copyright.txt for details.
+
+=========================================================================auto=*/
+
+#ifndef __vtkMRMLNodePropertyMacros_h
+#define __vtkMRMLNodePropertyMacros_h
+
+
```

Comment:
would be nice to have the macros documented in doxygen

Reply:
And mention that they have been designed to be used in method `vtkMRMLNode::Copy(), ... `

Reply:
I agree with @jcfr, I don't see what macros here do better than functions. 

```
ostream& vtkMRMLWriteXMLBoolean(ostream & xmlWriteOutputStream , const char* attributeName, bool propertyName)
{
  xmlWriteOutputStream << " "<< attributeName <<  "=\"" << propertyName ? "true" : "false" << "\"";
}
...

vtkMRML***Node::WriteXML(ostream& of, int nIndent)
{
...
   vtkMRMLWriteXMLBoolean(of, "layoutLabel", this->GetLayoutLabel());
...
}
```

Templated functions could also be used up to some extend.

Reply:
There are several reasons why macros more appropriate choice for this:

1. Simpler and more consistent syntax. Have a look at the current syntax and try to reproduce it without using macros.

2. Greater flexibility. We can create block with local variables, use the same argument as string and a method name, etc. allowing better performance, robustness, and keeping implementation simpler.

3. Future-proofing. Without changing existing MRML nodes, we can switch to use custom templated functions, serialization library, or (in a few years) standard C++ reflection.

Reply:
Hi @lassoan , 

Sounds good to me, I  just wanted to make sure we discuss it.

Could you add the three points written above in the commit message ?